nebula_decoders/include/nebula_decoders/nebula_decoders_common/point_filters/downsample_mask.hpp

```diff
@@ -0,0 +1,143 @@
+  uint8_t quantization_levels)
+{
+  if (in.get_width() != out.get_width() || in.get_height() != out.get_height()) {
+    std::stringstream ss;
+    ss << "Expected downsample mask of size "
+       << "(" << out.get_width() << ", " << out.get_height() << ")";
+    ss << ", got "
+       << "(" << in.get_width() << ", " << in.get_height() << ")";
+
+    throw std::runtime_error(ss.str());
+  }
+
+  uint32_t denominator = quantization_levels;
+
+  auto should_keep = [denominator](uint32_t numerator, uint32_t pos) {
+    for (uint32_t i = 0; i < numerator; ++i) {
+      auto dithered_pos =
+        static_cast<size_t>(std::round(denominator / static_cast<double>(numerator) * i));
+      if (dithered_pos == pos) return true;
+    }
+    return false;
+  };
+
+  for (size_t y = 0; y < out.get_height(); ++y) {
+    for (size_t x = 0; x < out.get_width(); ++x) {
+      const auto & pixel = in.get_pixel(x, y);
+      uint32_t numerator = static_cast<uint32_t>(pixel) * denominator / 255;
+      size_t pos = (x + y) % denominator;
+      bool keep = should_keep(numerator, pos);
+      out.set_pixel(x, y, keep * 255);
+    }
+  }
+}
+
+}  // namespace impl
+
+class DownsampleMaskFilter
+{
+  static const uint8_t g_quantization_levels = 10;
+
+public:
+  DownsampleMaskFilter(
+    const std::string & filename, AngleRange<int32_t> azimuth_range_mdeg,
+    uint32_t azimuth_peak_resolution_mdeg, size_t n_channels,
+    const std::shared_ptr<loggers::Logger> & logger, bool export_dithered_mask = false)
+  : azimuth_range_{deg2rad(azimuth_range_mdeg.min / 1000.), deg2rad(azimuth_range_mdeg.max / 1000.)}
+  {
+    png::image<png::gray_pixel> factors(filename);
+
+    size_t mask_cols = azimuth_range_mdeg.extent() / azimuth_peak_resolution_mdeg;
+    size_t mask_rows = n_channels;
+
+    png::image<png::gray_pixel> dithered(mask_cols, mask_rows);
+    impl::dither(factors, dithered, g_quantization_levels);
+
+    mask_ = Eigen::MatrixX<uint8_t>(mask_rows, mask_cols);
+
+    for (size_t y = 0; y < dithered.get_height(); ++y) {
+      for (size_t x = 0; x < dithered.get_width(); ++x) {
+        mask_.coeffRef(static_cast<int32_t>(y), static_cast<int32_t>(x)) = dithered.get_pixel(x, y);
+      }
+    }
+
+    if (export_dithered_mask) {
+      std::filesystem::path out_path{filename};
+      out_path = out_path.replace_filename(
+        out_path.stem().string() + "_dithered" + out_path.extension().string());
+
+      try {
+        dithered.write(out_path);
+        logger->info("Wrote dithered mask to " + out_path.native());
+      } catch (const png::std_error & e) {
+        logger->warn("Could not write " + out_path.native() + ": " + e.what());
+      }
+    }
+  }
+
+  bool excluded(const NebulaPoint & point)
+  {
+    double azi_normalized = (point.azimuth - azimuth_range_.min) / azimuth_range_.extent();
+
+    auto x = static_cast<ssize_t>(std::round(azi_normalized * static_cast<double>(mask_.cols())));
```

Comment:
Might be missing something, but I don't see why a <ssize_t> is used here when a <size_t> and any required clamping would be safer (as x should never be negative).

Reply:
`azi_normalized` can be smaller than 0, making the rounded result also smaller than 0.
I personally find it cleaner to not rely on the negative value wrapping around to a huge positive one.
I am checking for negativity in `x_out_of_bounds` before I access the mask, so this is safe even though `x` is a `ssize_t`.